### PR TITLE
29881 add ability to attach disk

### PR DESCRIPTION
**Pull request by SmallEmoLover**

### CODE REVIEW

File: core/services/terraform_service.rb [hunk 1]
```diff
@@ -119,6 +119,33 @@ def self.nodes_to_resources(nodes, resource_type)
...
+  end
+
+  # Determines if the specified node have attached disk resource
+  #
+  # @param nodes [String] name of the node
+  # @param logger [Logger] logger to log into
+  # @param path [String] path to the directory with terraform configuration
+  # @return [Boolean] Does the node has additional disk
+  def self.node_has_disk_resource?(node, logger, path = Dir.pwd)
+    ShellCommands.run_command_in_dir(
```
Comment (amvasilyev) [inline core/services/terraform_service.rb]: Why not just open file with ruby and search for the specified string?

File: core/commands/partials/terraform_configurator.rb [hunk 2]
```diff
@@ -71,7 +84,11 @@ def up_machines(nodes)
       @attempts.times do |attempt|
         @ui.info("Up nodes #{nodes}. Attempt #{attempt + 1}.")
         destroy_nodes(target_nodes.keys) if @recreate_nodes || attempt.positive?
-        result = TerraformService.apply(target_nodes.values, @ui, @config.path)
+        resources_to_bring = target_nodes.values
+        if @config.provider == 'aws'
+          resources_to_bring += TerraformService.additional_disk_resources(target_nodes.keys, @ui, @config.path)
```
Comment (amvasilyev) [inline core/commands/partials/terraform_configurator.rb]: The up_machines is the wrong location to calculate which resources are needed to be brought up. This method coordinates the whole up process. This logic must be moved to another places.

File: core/services/terraform_service.rb (hunk 1 above)
Comment (amvasilyev) [reply core/services/terraform_service.rb]: You call this method FOR EACH node that needs to be brought up. This is clearly the wrong direction as the Configuration class already provides means to access the template.

File: core/commands/partials/terraform_configurator.rb [hunk 3]
```diff
@@ -71,7 +84,11 @@ def up_machines(nodes)
       @attempts.times do |attempt|
         @ui.info("Up nodes #{nodes}. Attempt #{attempt + 1}.")
         destroy_nodes(target_nodes.keys) if @recreate_nodes || attempt.positive?
-        result = TerraformService.apply(target_nodes.values, @ui, @config.path)
+        resources_to_bring = target_nodes.values
+        if @config.provider == 'aws'
```
Comment (amvasilyev) [inline core/commands/partials/terraform_configurator.rb]: Currently only AWS is supported by this?

File: core/commands/partials/terraform_configuration_generator.rb [hunk 4]
```diff
@@ -138,7 +138,7 @@ def make_node_params(node, box_params)
   def need_attached_disk?(node)
     ConfigurationGenerator.parse_products_info(node).map do |product|
       product['name']
-    end.include?('clustrix')
+    end.include?('clustrix') || node[1]['attach_disk'] == 'true'
```
Comment (amvasilyev) [inline core/commands/partials/terraform_configuration_generator.rb]: This meta-information may be put into the corresponding product description. This is not priority direction.

File: core/commands/partials/terraform_configurator.rb (hunk 3 above)
Comment (SmallEmoLover) [reply core/commands/partials/terraform_configurator.rb]: @amvasilyev 
Disks for GCP machines are applied automatically when bringing up machine instance.
We need to search additional resources only for AWS disks.

File: core/services/terraform_service.rb [hunk 5]
```diff
@@ -119,6 +119,22 @@ def self.nodes_to_resources(nodes, resource_type)
     nodes.map { |node| [node, "#{resource_type}.#{node}"] }.to_h
   end
 
+  # Searches additional disk resources declared in the terraform config file
+  #
+  # @param path [String] path to the directory with terraform configuration
+  # @return [Hash] Hash with node name keys and resources names values
+  # in format 'aws_volume_attachment.node-disk-attachment'
+  def self.additional_disk_resources(path)
+    configuration_path = "#{path}/#{TerraformConfigurationGenerator::CONFIGURATION_FILE_NAME}"
```
Comment (amvasilyev) [inline core/services/terraform_service.rb]: For each node that has the attached disk, there is a "NAME-disk-attachment" resource. Why do we need to parse the terraform configuration file in order to determine that? We can determine that requirement solely from the template file.

_Another point - why do you tie this logic to the AWS only? It only depends on the requirement of the disk attachment._ This part was answered.

**Pull request merged**